### DISCUSSION TITLE
UDP datasource?

### DISCUSSION THREAD

**Discussion by OrenShm** [Kittyfisto/Tailviewer#253] [Ideas]
Hi,

First of all - I really like this logger, thanks for the great tool! 👍
Our application logger is sending the payloads via UDP (from external device).
Currently for viewing the logs we have to open wireshark, copy the payloads and paste it into a file in order to view the logs.

I was wondering if there's any option to add UDP as data source? :)

Thanks a lot!
Oren.

**Comment by Kittyfisto** [comment]
@OrenShm Thanks, I'm happy to receive feedback and am glad you like it :)

Unfortunately, there is no option to "out of the box" display UDP packets (or their contents) in tailviewer.

One option is to write a simple windows application to receive those UDP packets and to stream their contents into a simple text file and then have Tailviewer tail this file.  

Another option is to write a plugin for tailviewer. There are interfaces to add custom data sources for tailviewer, and there are a few examples as well as some documentation, but there's nothing out of the box just yet.

Maybe you can explain to me what format you are using (are the log messages simply ASCII text in the payload of the UDP datagram?) and then we can figure out a way to get it into tailviewer. So as long as it's not too application specific (e.g. if multiple people could use it)  I'll happily add it to tailviewer for everyone. If it's highly specific for one application then I'll still be able to help, but most probably in the form of a custom-tailored plugin.

Here are some questions from the top of my head:

1. What's the form of the message?
2. What encoding is used? ASCII, UTF-8, etc..?
3. Any form of data integrity on top of UDP? (Duplicates, missing messages, corrupt messages, etc...)
4. Can you provide some example (or real, if allowed) log messages?
5. How much data are we talking about here? E.g. is it meant to be used in a "monitoring" scenario where old messages should be "purged" after some time as to not have it overflow?
6. Should the received messages be persisted in a file as well?
7. Could the plugin open a simple listening socket on the target port (e.g. does the application which usually receives said datagrams allow sharing) or is it necessary to use winpcap to intercept those datagrams?

**Reply by OrenShm** [reply]
Thanks Simon!
1. It's a really simple logger which for every "LOG_INFO("Hello world");" it would send the string "<timestamp> INFO: Hello world" via UDP packet.
2. Just string message without any encoding.
3. We don't enforce any data integrity over the UDP.
4. The project is developed within inside our organization which is not linked directly to the internet so nope.. 
but the output is just normal form of log. (not so far from the original TailViewer log)
5. We are talking about up to 10MB log data over few hours so I don't think we'll have any overflow issue.
(It would just be easier to view the log instead of copying it via the wireshark).
6. I don't really need persistent, just "real time" logger which listen to the UDP port and extracts the data.
I don't mind it would also be saved into a file.
7. We could just create an app like you suggested which listen to dedicated UDP port..

Do you think the Socket listening is app-specific? if so we will create some side application as you suggested or add a plugin

**Reply by Kittyfisto** [reply]
@OrenShm From what you've described it could be used by multiple people (as in, this doesn't look terribly application-specific).  
 
I can envision this feature as part of Tailviewer, but I'd have to do a little more ground-work before this can be achieved: There's no practical way for plugins to add a proper UI for example to allow a user to input certain values as part of a data source's configuration, which in your case would be the address and port on which to listen for packets.  

If I were to develop this feature I'd probably need some feedback from you in the process: It's unlikely that the first iteration will be good enough. I have some free time coming up over christmas and may have a first prototype running until the end of this year. Please let me know if this agrees with you and we can hash out some more details.

**Reply by OrenShm** [reply]
Sounds great! 
I could give you that feedback :)

Thanks a lot..

**Reply by Kittyfisto** [reply]
@OrenShm Sorry for the long delay, I had to take an extended break over the holidays, but now I'm back. I just wanted to let you know that I still want to follow through with this. However I have to do some additional work before I can get something running.

**Reply by OrenShm** [reply]
It's okay. thanks for the update :) 
eager to see that coming 👍

**Reply by OrenShm** [reply]
Hi,
There's been a while, any estimation if and when that feature would be added?
Can I help somehow?

Thanks,
Oren.
